### ISSUE TITLE
Segfault in BMV2 compiler

### ISSUE THREAD

**Issue by cole-barefoot** [p4lang/p4c#376]
The following code causes the BMV2 compiler to segfault:

```p4
#include "v1model.p4"                                                                                                                                 

struct headers_t { bit<8> h; }                                                                                                                        
struct meta_t { bit<8> unused; }                                                                                                                      

parser p(packet_in b, out headers_t hdrs, inout meta_t m, inout standard_metadata_t meta)                                                             
{                                                                                                                                                     
    state start {                                                                                                                                     
        b.extract(hdrs);                                                                                                                              
        transition accept;                                                                                                                            
    }                                                                                                                                                 
}                                                                                                                                                     

control i(inout headers_t hdrs, inout meta_t m, inout standard_metadata_t meta) {                                                                     
    action a() { meta.egress_spec = 1; }                                                                                                              
    apply { a(); }                                                                                                                                    
}                                                                                                                                                     

control e(inout headers_t hdrs, inout meta_t m, inout standard_metadata_t meta) {                                                                     
    apply { }                                                                                                                                         
}                                                                                                                                                     

control d1(packet_out packet, in headers_t hdrs) {                                                                                                    
    apply { packet.emit(hdrs); }                                                                                                                      
}                                                                                                                                                                                                                                                                                      

control vc(in headers_t hdrs, inout meta_t meta) { apply { } }                                                                                        
control cc(inout headers_t hdrs, inout meta_t meta) { apply { } }                                                                                     
                                                                                                                                                      
V1Switch(p(), vc(), i(), e(), cc(), d1()) main;
```

**Comment by mbudiu-vmw** [comment]
I have a fix, but I am waiting for PR #367 to be merged.
The program will fail to compile with the following error:

```
x.p4(25): error: hdrs: extract only accepts arguments with header types, not struct headers_t
        b.extract(hdrs);
                  ^^^^
x.p4(19)
struct headers_t { bit<8> h; }
       ^^^^^^^^^
x.p4(40): error: hdrs: emit only supports header and stack arguments, not struct headers_t
    apply { packet.emit(hdrs); }
                        ^^^^
x.p4(19)
struct headers_t { bit<8> h; }
       ^^^^^^^^^
```

**Comment by mbudiu-vmw** [comment]
This issue should have been solved.
